pallets/pool-registry/src/lib.rs

```
@@ -0,0 +1,392 @@
+		TrancheTokenNameTooLong,
+		/// The given tranche symbol name exceeds the length limit
+		TrancheSymbolNameTooLong,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a new pool
+		///
+		/// Initialise a new pool with the given ID and tranche
+		/// configuration. Tranche 0 is the equity tranche, and must
+		/// have zero interest and a zero risk buffer.
+		///
+		/// The minimum epoch length, and maximum NAV age will be
+		/// set to chain-wide defaults. They can be updated
+		/// with a call to `update`.
+		///
+		/// The caller will be given the `PoolAdmin` role for
+		/// the created pool. Additional administrators can be
+		/// added with the Permissions pallet.
+		///
+		/// Returns an error if the requested pool ID is already in
+		/// use, or if the tranche configuration cannot be used.
+		#[pallet::weight(T::WeightInfo::create(tranche_inputs.len().try_into().unwrap_or(u32::MAX)))]
+		#[transactional]
+		pub fn create(
+			origin: OriginFor<T>,
+			admin: T::AccountId,
+			pool_id: T::PoolId,
+			tranche_inputs: Vec<
+				TrancheInput<T::Rate, T::MaxTokenNameLength, T::MaxTokenSymbolLength>,
+			>,
+			currency: T::CurrencyId,
+			max_reserve: T::Balance,
+			metadata: Option<Vec<u8>>,
+		) -> DispatchResult {
+			T::PoolCreateOrigin::ensure_origin(origin.clone())?;
+
+			// First we take a deposit.
+			// If we are coming from a signed origin, we take
+			// the deposit from them
+			// If we are coming from some internal origin
+			// (Democracy, Council, etc.) we assume that the
+			// parameters are vetted somehow and rely on the
+			// admin as our depositor.
+			let depositor = ensure_signed(origin).unwrap_or(admin.clone());
+
+			match T::ModifyPool::create(
+				admin.clone(),
+				depositor,
+				pool_id,
+				tranche_inputs,
+				currency,
+				max_reserve,
+				metadata,
+			) {
+				Ok(_) => {
+					Self::deposit_event(Event::Created { pool_id, admin });
+					Ok(())
+				}
+				Err(e) => Err(e),
+			}
```

`map(|_| Self::deposit_event(Event::Created{pool_id, admin}))` instead of match. Or was this not possible?

---

Hm... doesn't this swallow the `Error`?

---

No. See https://doc.rust-lang.org/src/core/result.rs.html#767

---

Oh sweet!